### DISCUSSION TITLE
Trying to model evaporation

### DISCUSSION THREAD

**Discussion by L278** [PMEAL/OpenPNM#1926] [Q&A]
Hi, I am trying to model evaporation, I've tried to modify the ordinary percolation and porosimetry algorithms but I was unsuccessful. One thing I'm struggling with is how to make the pores and throats be saturated with water at the start with air invading due to pressure changes. I've come across something called 'Drainage' in some documentation, where an invading and defending phase can be chosen, however, there doesn't seem to be an actual drainage algorithm. Any guidance/ suggestions are greatly appreciated.

**Comment by jgostick** [comment]
We don't have any explicit algorithms for evaporation.  It's quite a complicated process actually.  We do have drainage algorithms, both OP and IP can do drainage.  Drainage refers to the invasion of a non-wetting phase into the domain, so in the case of evaporation, drainage means air invading into the domain in the form of a drying front.  Water vapor then exists the domain via the air filled pores.  The reason evaporation/drying is hard is that the front moves and this is tied to the diffusion rate of water vapor out of the system. There is also air-water surface area to think about, which probably needs to include wetting films of water.  Also heat transfer is probably important, so things are non-isothermal.  And lastly there is phase partitioning/vapor pressure think about.  Basically, you don't see many pore network models of evaporation for a reason...it's very hard to do right.  It's not impossible, but it's not just a 'plug and play' type of simulation.

**Reply by L278** [reply]
I think the drainage algorithm will do, for now, thank you. I also have a question about it, as I keep getting an error:
```python
import openpnm as op
import matplotlib.pyplot as plt
import numpy as np
np.random.seed(10)

ws = op.Workspace()
ws.settings["loglevel"] = 40

pn = op.network.Cubic(shape=[10, 10, 10], spacing=0.0001)
geom = op.geometry.StickAndBall(network=pn, pores=pn.Ps, throats=pn.Ts)

Water = op.phases.Water(network=pn)
Air = op.phases.Air(network=pn)
phys = op.physics.Standard(network=pn, phase=Water, geometry=geom)
**MIP = op.algorithms.Drainage(network=pn)**
MIP.setup(invading_phase=Air, defending_phase=Water)
MIP.set_inlets(pores=pn.pores(['top', 'bottom']))
MIP.run()
```
I keep getting the error that module 'openpnm.algorithms' has no attribute 'drainage'
I'm not sure what the correct term for drainage is as I can't seem to find documentation on it.

**Reply by jgostick** [reply]
It's not called Drainage.  It's called OrdinaryPercolation or InvasionPercolation.   You'll want to look at the examples [here](https://github.com/PMEAL/OpenPNM/tree/dev/examples/notebooks/algorithms/percolation)

**Reply by L278** [reply]
Thank you and thanks for answering all my questions! I know you said there isn't a way to model evaporation by plugging and playing but is there a way to make the sample start fully saturated. To my understanding in InvasionPercolation and OrdinaryPercolation the sample starts off dry and then water percolates through pores in steps according to the pressure.

**Reply by jgostick** [reply]
We did a [paper a few years ago](https://doi.org/10.1007/s10800-017-1126-6) that used a rough procedure that you can try.

You start with a water filled domain, and compute the water-vapor diffusion rate from the top of the domain. You also know which pore is the easier to invade with air (air is non-wetting).  Given the volume of this pore and the evaporation rate, you can find how long it will take to dry the first pore.  You then set this pore to air filled, and recompute the diffusion rate, which now has additional area.  You can then repeat until the domain is dry.

The BIGGEST problem is dealing with disconnected water blobs.  These will evaporate too, so now you need to monitor the rate of water leaving each blob, and doing the above trick on each separately.

Like I said...it's messy.  My collaborator Pablo Garcia-Salaberri has developed a ['continuum-pnm' hybrid platform](https://doi.org/10.1016/j.ijheatmasstransfer.2020.120824) using Ansys so that he could leverage the coupled heat and mass transfer while still having the pnm description of the two-phase transport.  I'm not sure if he's got to the point of evaporation yet though.

**Reply by L278** [reply]
Thank you for sharing the papers with me, they were quite interesting.
